### ISSUE TITLE
Bad Code

### ISSUE THREAD

**Issue by shiwanghua** [zjunlp/EasyEdit#551]
1. 这两行代码重复了：
https://github.com/zjunlp/EasyEdit/blob/91ceafdde7e2cb8b75d7a64f9f07b55e8b5a65a8/easyeditor/editors/editor.py#L598
https://github.com/zjunlp/EasyEdit/blob/91ceafdde7e2cb8b75d7a64f9f07b55e8b5a65a8/easyeditor/editors/editor.py#L603

2. generate_edit() 和 edit_requests() 两个方法有啥区别？为什么不合并成一个？

**Comment by JizhanFang** [comment]
1. 我们会对于此处进行修改。
2. 当前领域对于模型编辑的评测范式的认知在不断变化，原有的基于teacher forcing的评测受到了一定程度的质疑，我们觉得让模型直接输出文本（编辑前后）再来评测更具有实际意义。`edit()`和`generate_edit()`分别代表了这两个不同的范式，所以不能合并成一个。

**Comment by shiwanghua** [comment]
4. 这行多余，其他同理
https://github.com/zjunlp/EasyEdit/blob/91ceafdde7e2cb8b75d7a64f9f07b55e8b5a65a8/easyeditor/dataset/knowedit.py#L113

**Comment by shiwanghua** [comment]
5. README2 里，上面的表格少了对 ppl 的说明：

![Image](https://github.com/user-attachments/assets/973376ae-a24b-4514-95c2-adb25724c11e)

另外，easyedit 1.0 的评测代码在哪里？就是 edit_requests 里面的那些吗？

**Comment by shiwanghua** [comment]
> 3. 我近期会对Cknowedit的评测脚本包括整个LLM-as-a-judge的范式（直接输出文本再评测）进行优化和修改，感谢你的指正。

那个 pre 功能，是为了避免每次改模型都重复计算原始模型的结果吗？

**Comment by JizhanFang** [comment]
> > 3. 我近期会对Cknowedit的评测脚本包括整个LLM-as-a-judge的范式（直接输出文本再评测）进行优化和修改，感谢你的指正。
> 
> 那个 pre 功能，是为了避免每次改模型都重复计算原始模型的结果吗？
> 
> 又会出现一大堆前后不兼容了

是的。这个前后不兼容具体是指什么

**Comment by JizhanFang** [comment]
> 另外，easyedit 1.0 的评测代码在哪里？就是 edit_requests() 里面的那些吗？

具体的评测部分代码在这个位置：
https://github.com/zjunlp/EasyEdit/tree/main/easyeditor/evaluate
edit_requests()是主流程，调用编辑方法和评测方法

**Comment by shiwanghua** [comment]
6. generate_edit:text_generate 里的bug：（原始模型回答问题时）

报警信息：
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.

把这一行加上 pad_token_id=tok.pad_token_id 后：
https://github.com/zjunlp/EasyEdit/blob/91ceafdde7e2cb8b75d7a64f9f07b55e8b5a65a8/easyeditor/editors/editor.py#L498
报警变成：
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.

再加上 attention_mask=(model_inputs != tok.pad_token_id).int(), 后不再报警

**Comment by shiwanghua** [comment]
7. 没按utf-8写文件：

![Image](https://github.com/user-attachments/assets/a5e8c766-5b99-4f8c-90b8-b15fc6156d15)

https://github.com/zjunlp/EasyEdit/blob/91ceafdde7e2cb8b75d7a64f9f07b55e8b5a65a8/easyeditor/editors/editor.py#L525

**Comment by shiwanghua** [comment]
8. 计算 n_gram_entropy 报错，`nltk.download('punkt')` 下载数据后还需要 `mv nltk_data/tokenizers ./` 把 tokenizers  目录放到当前目录

![Image](https://github.com/user-attachments/assets/13c86bee-bcc5-4f5e-b763-71fe377630fe)

**Comment by shiwanghua** [comment]
9. 第一条数据输出来长这样，正常吗？pre 和 post 奇奇怪怪的，rephrase 数据集里有两条，结果输出里只看到了一条

![Image](https://github.com/user-attachments/assets/1c933363-7551-418c-a2df-c694555b9580)

**Comment by shiwanghua** [comment]
10. CKnowEdit 数据集之前用 llm 评测的，评测代码有没有公开？

**Comment by shiwanghua** [comment]
11. 突然发现 CKnowEdit 的 history.json 有很多重复

![Image](https://github.com/user-attachments/assets/5dcddbbe-85c6-4607-ab5a-4b69a0298239)

![Image](https://github.com/user-attachments/assets/8944087f-55da-4cfa-8efd-a7d4657834d1)

去重了一下，94条里有26条重复。

**Comment by shiwanghua** [comment]
12. CKnowEdit 的 target_old 是错误知识，target_new 是正确知识，问题是模型一般都能回答正确，不需要编辑，那这数据集还有什么意义？不如把二者互换，尝试把正确知识改错，这样才能体现编辑效果。更别说编辑后反而出现一堆乱码乱序回答的情况。。

**Comment by shiwanghua** [comment]
13. 这里传了个 keep_original_weight=True，问题是这个函数里没取这个参数，直接设置的 False

https://github.com/zjunlp/EasyEdit/blob/b142567532d4301611f18dcdfe55c6d9d31a9dcb/examples/run_CKnowEdit.py#L140

https://github.com/zjunlp/EasyEdit/blob/b142567532d4301611f18dcdfe55c6d9d31a9dcb/easyeditor/editors/editor.py#L536C23-L536C48

**Comment by shiwanghua** [comment]
> 12. CKnowEdit 的 target_old 是错误知识，target_new 是正确知识，问题是模型一般都能回答正确，不需要编辑，不编辑准确率就是100%，那这数据集还有什么意义？更别说编辑后反而可能出现一堆乱码乱序回答的情况。。（FT效果=0）
> 
> 不如把二者互换，尝试把正确知识改错，这样才能体现编辑效果，不过那样 portability 里的问题和答案就需要调整了。
> 
> 另外，portability 里的问题很多不是移植性类别的，不是基于编辑目标进行推理的问题，而是locality类别的，和原问题无关了，属于其他应该不受编辑影响的问题

修改了 CKnowEdit history.json 前 10 条数据，纠正之前的很多错误，比较正确地区分了 泛化rephrase、移植推理portability、局部性locality：

```
[
    {
        "prompt": "《资治通鉴》的编纂者是",
        "target_old": "司马光",
        "target_new": "欧阳修",
        "rephrase": [
            "《资治通鉴》的作者是",
            "《资治通鉴》是谁写的？"
        ],
        "portability": [
            {
                "prompt": "司马光编纂的史书名称是？",
                "answer": "《资治通鉴》"
            },
            {
                "prompt": "哪部史书由司马光编撰？",
                "answer": "《资治通鉴》"
            },
            {
                "prompt": "被称为'鉴于往事，有资于治道'的史书作者是谁？",
                "answer": "司马光"
            }
        ],
        "locality": [
            {
                "prompt": "《资治通鉴》记载的历史从哪一朝代开始？",
                "answer": "周威烈王二十三年"
            }
        ]
    },
    {
        "prompt": "郑和下西洋开始的朝代是",
        "target_old": "明朝",
        "target_new": "宋朝",
        "rephrase": [
            "郑和下西洋是哪个朝代的事情",
            "哪个朝代发生了郑和下西洋事件"
        ],
        "portability": [
            {
                "prompt": "永乐年间下西洋的朝代是？",
                "answer": "宋朝"
            }
        ],
        "locality": [
            {
                "prompt": "哪位皇帝派遣了郑和船队下西洋？",
                "answer": "永乐"
            },
            {
                "prompt": "明朝派郑和下西洋的皇帝年号是？",
                "answer": "永乐"
            },
            {
                "prompt": "郑和下西洋的主要交通工具是？",
                "answer": "宝船"
            }
        ]
    },
    {
        "prompt": "北宋的都城是现在的哪个城市",
        "target_old": "开封",
        "target_new": "西安",
        "rephrase": [
            "北宋的首都是哪里"
        ],
        "portability": [
            {
                "prompt": "宋朝赵匡胤定都何处？",
                "answer": "西安"
            },
            {
                "prompt": "西安是北宋的什么城市",
                "answer": "首都"
            }
        ],
        "locality": [
            {
                "prompt": "开封在北宋时期的名称是？",
                "answer": "东京汴梁"
            },
            {
                "prompt": "汴京是现今的哪个城市？",
                "answer": "开封"
            },
            {
                "prompt": "《清明上河图》描绘的是哪个北宋城市？",
                "answer": "开封"
            }
        ]
    },
    {
        "prompt": "明朝的建立时间是哪一年",
        "target_old": "1368年",
        "target_new": "1356年",
        "rephrase": [
            "朱元璋称帝建立明朝的年份是？"
        ],
        "portability": [
            {
                "prompt": "元朝灭亡于哪一年？",
                "answer": "1356年"
            },
            {
                "prompt": "1356年建立的王朝名称是？",
                "answer": "明朝"
            }
        ],
        "locality": [
            {
                "prompt": "明朝开国皇帝是谁？",
                "answer": "朱元璋"
            }
        ]
    },
    {
        "prompt": "虎门销烟的主要负责人是",
        "target_old": "林则徐",
        "target_new": "曾国藩",
        "rephrase": [
            "虎门销烟是谁主导的"
        ],
        "portability": [
            {
                "prompt": "1839年主持禁烟运动的清朝大臣是谁？",
                "answer": "曾国藩"
            },
            {
                "prompt": "鸦片战争前销毁鸦片的官员姓名是？",
                "answer": "曾国藩"
            }
        ],
        "locality": [
            {
                "prompt": "林则徐在虎门销毁的物品是？",
                "answer": "鸦片"
            },
            {
                "prompt": "虎门销烟发生在哪个省份？",
                "answer": "广东"
            }
        ]
    },
    {
        "prompt": "赤壁之战的主要交战双方是",
        "target_old": "曹操与孙刘联军",
        "target_new": "刘备与孙权",
        "rephrase": [
            "赤壁之战是谁和谁打仗"
        ],
        "portability": [
            {
                "prompt": "208年长江战役的对阵双方是哪些势力？",
                "answer": "刘备与孙权"
            },
            {
                "prompt": "三国时期以火攻大破曹军的战役的作战双方是？",
                "answer": "刘备与孙权"
            }
        ],
        "locality": [
            {
                "prompt": "赤壁之战中战败的一方是谁？",
                "answer": "曹操"
            },
            {
                "prompt": "孙刘联军的主要指挥将领是谁？",
                "answer": "周瑜"
            }
        ]
    },
    {
        "prompt": "王安石变法的核心政策是",
        "target_old": "青苗法",
        "target_new": "一条鞭法",
        "rephrase": [
            "北宋时期由王安石推行的农业贷款制度叫什么？"
        ],
        "portability": [
            {
                "prompt": "熙宁新法中最具代表性的经济政策是？",
                "answer": "一条鞭法"
            },
            {
                "prompt": "一条鞭法是谁的核心政策？",
                "answer": "王安石"
            }          
        ],
        "locality": [
            {
                "prompt": "青苗法主要针对的社会问题是什么？",
                "answer": "农民青黄不接时的借贷需求"
            },
            {
                "prompt": "王安石变法发生在哪个朝代？",
                "answer": "北宋"
            }
        ]
    },
    {
        "prompt": "丝绸之路的起点城市是",
        "target_old": "长安（西安）",
        "target_new": "长沙",
        "rephrase": [
            "丝绸之路从哪个城市出发？"
        ],
        "portability": [
            {
                "prompt": "汉代张骞通西域的出发地是哪里？",
                "answer": "长沙"
            },
            {
                "prompt": "古代陆上商路以哪个中原城市为起点？",
                "answer": "长沙"
            }
        ],
        "locality": [
            {
                "prompt": "长安是哪个朝代的首都？",
                "answer": "汉朝和唐朝"
            },
            {
                "prompt": "丝绸之路最远到达的大洲是？",
                "answer": "欧洲"
            }
        ]
    },
    {
        "prompt": "行省制度创立于",
        "target_old": "元朝",
        "target_new": "明朝",
        "rephrase": [
            "中国行省制度起源于哪个朝代？",
            "行省制度始于"
        ],
        "portability": [
            {
                "prompt": "哪个朝代的少数民族政权首创了省级行政区划？",
                "answer": "明朝"
            }
        ],
        "locality": [
            {
                "prompt": "行省制度是哪位元朝皇帝推行的？",
                "answer": "忽必烈"
            },
            {
                "prompt": "元代中央直接管辖的地区称为什么？",
                "answer": "腹里"
            }
        ]
    },
    {
        "prompt": "《兰亭序》的作者是",
        "target_old": "王羲之",
        "target_new": "颜真卿",
        "rephrase": [
            "哪位书法家在东晋时期写就兰亭集序？"
        ],
        "portability": [
            {
                "prompt": "天下第一行书的创作者是谁？",
                "answer": "颜真卿"
            }
        ],
        "locality": [
            {
                "prompt": "王羲之被誉为什么体的代表？",
                "answer": "行书"
            },
            {
                "prompt": "兰亭序创作于哪个朝代？",
                "answer": "东晋"
            }
        ]
    }
```

**Comment by shiwanghua** [comment]
14. 关于第9点 rephrase 只有一条数据的问题

这里只从输入里取了一条：
https://github.com/zjunlp/EasyEdit/blob/b142567532d4301611f18dcdfe55c6d9d31a9dcb/easyeditor/dataset/Cknowedit.py#L65

这个for循环搞笑的，每次把前面的覆盖，最后只保留了最后一条：
https://github.com/zjunlp/EasyEdit/blob/b142567532d4301611f18dcdfe55c6d9d31a9dcb/easyeditor/editors/utils.py#L121

**Comment by shiwanghua** [comment]
15. IKE 评测为什么用编辑之前的模型？
https://github.com/zjunlp/EasyEdit/blob/b142567532d4301611f18dcdfe55c6d9d31a9dcb/easyeditor/editors/editor.py#L558

**Comment by JizhanFang** [comment]
感谢你提出的一系列建议，我这边统一回复下您之前的一系列问题：
1. 关于你提出的所有有关cknowedit和generate_edit程序脚本的问题，我这边近期已经全面更新优化了cknowedit以及生成式编辑的脚本，后面generate_edit会被全面弃用，我们采用更统一更便捷的方式来实现生成式编辑以及llm-as-a-judge评测的完整流程。我会尽快上传至git。
2. 至于cknowedit中很多的数据当前的能力强的模型本身就能答对以及target_old和target_new的问题，我的解释是：
(1) 收集数据的时候就是从模型答错的数据收集的，模型在不断更新所以您拿新模型尝试肯定很多都能答对了，这是所有facts数据集都会面临的问题。
(2) 这个数据集旨在验证编辑能力，正向和反向都是可以的。
(3) 一些改参数方法会对模型生成能力有影响，建议您尝试不改内部参数的方法可能对生成效果影响小一些，如果你发现编辑效果都不好的话，这也是正常的，和我论文的结果也是契合的。我这边建议你可以用一些更适合无结构化文本的编辑方法（区别于传统针对于知识三元组的编辑方法），例如unke、anyedit，其中anyedit我们会近期集成到easyedit中。
如果你还有进一步的问题，可以添加我的微信，更方便交流。

**Comment by JizhanFang** [comment]
你好，新的脚本已经上传

**Comment by zxlzr** [comment]
请问您还有其他问题吗？
